implementations/elixir/ockam/ockam_hub/lib/hub/service/token_lease_manager.ex

```diff
@@ -1,130 +1,242 @@
+    else
+      error -> error
+    end
+  end
 
-    storage_service_name =
-      case Keyword.fetch(options, :storage_service) do
-        {:ok, storage_service_name} -> storage_service_name
-        _other -> @default_storate_service
-      end
+  ### for exposing service ##
 
-    state = %{
-      cloud_service: get_cloud_service(cloud_service_name),
-      storage_system: get_storage_service(storage_service_name)
-    }
+  @impl true
+  def handle_message(%{payload: payload, return_route: return_route}, state) do
+    reply_payload = process_request(payload, state)
 
-    # TODO: initialization for all expirations
-    # Process.send(self(), :set_all_expiration, [])
+    Router.route(%{
+      payload: reply_payload,
+      onward_route: return_route,
+      return_route: [state.address]
+    })
 
     {:ok, state}
   end
 
-  @impl true
-  def handle_info({:create, options}, state) do
-    case create(state[:cloud_service], options) do
-      {:ok, _lease} -> {:noreply, state}
-      error -> error
-    end
-  end
+  ### for self use ##
 
-  def handle_info({:revoke, token_id}, state) do
-    case revoke(state[:cloud_service], token_id) do
-      :ok -> {:noreply, state}
-      error -> error
-    end
-  end
-
-  def handle_info({:set_expiration, token_id, expiration}, state) do
-    set_expiration(self(), token_id, expiration)
-    {:noreply, state}
+  def handle_message({:set_ttl, token_id, ttl}, state) do
+    set_ttl(self(), token_id, ttl)
+    {:ok, state}
   end
 
-  def handle_info(:set_all_expiration, state) do
+  def handle_message(:set_all_ttl, %{storage_service: storage_service} = state) do
+    Logger.info("setting all ttls")
     lease_manager_pid = self()
 
     Task.start(fn ->
-      Enum.map(get_all_leases(), fn lease ->
-        set_expiration(lease_manager_pid, lease[:token_id], lease[:expiration])
+      {:ok, leases} = storage_service.get_all()
+
+      Enum.map(leases, fn lease ->
+        ttl = new_ttl(lease.ttl, lease.issued)
+        set_ttl(lease_manager_pid, lease.id, ttl)
       end)
     end)
 
-    {:noreply, state}
+    {:ok, state}
   end
 
-  def handle_info(_message, state) do
-    {:noreply, state}
+  def handle_message({:auto_revoke, token_id}, state) do
+    Logger.info("revoking #{token_id} token")
+    Task.start(fn -> do_revoke(state[:cloud_service], state[:storage_service], token_id) end)
+    {:ok, state}
   end
 
-  defp get_cloud_service(name) do
-    case name do
-      :influxdb -> Ockam.TokenLeaseManager.CloudServiceInfluxdb
-      _other -> Ockam.TokenLeaseManager.CloudServiceInfluxdb
-    end
+  defp process_request(received_payload, state) do
+    {_prefix, payload} = get_payload_prefix(received_payload)
+
+    response = process_encoded_payload(payload, state)
+
+    length =
+      case String.length(response) do
+        l when l > 128 and l < 255 ->
+          :binary.encode_unsigned(l) <> <<1>>
+
+        l ->
+          :binary.encode_unsigned(l)
+      end
+
+    length <> response
```

Comment:
I suspect that's a replacement for `:bare.encode(string, :string)`, is that so?

Reply:
I do not think so.

I need to encode in 16 bytes the length of the message which is an integer. Let me know if you see anything weird.

Reply:
You're trying to decode/encode a string compatible with Rust. Rust does BARE encoding with a `string` type. So let's use BARE in elixir as well.